sdk/testing/stepwise/environments/docker/environment.go

```diff
@@ -0,0 +1,851 @@
+	"github.com/hashicorp/go-cleanhttp"
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/vault/api"
+	"github.com/hashicorp/vault/internalshared/reloadutil"
+	"github.com/hashicorp/vault/sdk/helper/consts"
+	"github.com/hashicorp/vault/sdk/testing/stepwise"
+	"golang.org/x/net/http2"
+
+	docker "github.com/docker/docker/client"
+	uuid "github.com/hashicorp/go-uuid"
+)
+
+var _ stepwise.Environment = &DockerCluster{}
+
+// DockerCluster is used to managing the lifecycle of the test Vault cluster
+type DockerCluster struct {
+	// PluginName is the input from the test case
+	PluginName string
+	// ClusterName is a UUID name of the cluster.
+	ClusterName string
+
+	// MountOptions are a set of options for registering and mounting the plugin
+	MountOptions stepwise.MountOptions
+
+	RaftStorage   bool
+	CACertBytes   []byte
+	CACertPEM     []byte
+	CAKeyPEM      []byte
+	CACertPEMFile string
+	ID            string
+	ClusterName   string
+	RootCAs       *x509.CertPool
+	CACert        *x509.Certificate
+	CAKey         *ecdsa.PrivateKey
+	CleanupFunc   func()
+	SetupFunc     func()
+	ClusterNodes  []*DockerClusterNode
+
+	// networkID tracks the network ID of the created docker network
+	networkID string
+
+	barrierKeys  [][]byte
+	recoveryKeys [][]byte
+	tmpDir       string
+
+	clientAuthRequired bool
+	// the mountpath of the plugin under test
+	mountPath string
+	// rootToken is the initial root token created when the Vault cluster is
+	// created.
+	rootToken string
+}
+
+// Teardown stops all the containers.
+func (rc *DockerCluster) Teardown() error {
+	var result error
+	for _, node := range rc.ClusterNodes {
+		if err := node.Cleanup(); err != nil {
+			result = multierror.Append(result, err)
+		}
+	}
+
+	//clean up networks
+	if rc.networkID != "" {
+		cli, err := docker.NewClientWithOpts(docker.FromEnv, docker.WithVersion("1.40"))
+		if err != nil {
+			return err
+		}
+		if err := cli.NetworkRemove(context.Background(), rc.networkID); err != nil {
+			return err
+		}
+	}
+
+	return result
+}
+
+// MountPath returns the path that the plugin under test is mounted at. If a
+// MountPathPrefix was given, the mount path uses the prefix with a uuid
+// appended. The default is the given PluginName with a uuid suffix.
+func (dc *DockerCluster) MountPath() string {
+	if dc.mountPath != "" {
+		return dc.mountPath
+	}
+
+	uuidStr, _ := uuid.GenerateUUID()
+	prefix := dc.PluginName
+	if dc.MountOptions.MountPathPrefix != "" {
+		prefix = dc.MountOptions.MountPathPrefix
+	}
+
+	dc.mountPath = fmt.Sprintf("%s_%s", prefix, uuidStr)
+	if dc.MountOptions.PluginType == stepwise.PluginTypeCredential {
+		dc.mountPath = fmt.Sprintf("%s/%s", "auth", dc.mountPath)
+	}
+
+	return dc.mountPath
+}
+
+// RootToken returns the root token of the cluster, if set
+func (dc *DockerCluster) RootToken() string {
+	return dc.rootToken
+}
+
+// Name returns the name of this environment
+func (dc *DockerCluster) Name() string {
+	return "Docker"
```

Comment:
Not sure how this is used, but if it's for a more programatic comparison maybe lowercase names would be better?

Reply:
I lower cased the name in 113085c02. It's use is not apparent now but future plans may use it.